### ISSUE TITLE
Repeated calls to `cGroup.Consume` go to the same Kafka broker even if it's down

### ISSUE THREAD

**Issue by ZoabKapoor** [IBM/sarama#2952]
##### Description

Hi there! My colleague @PapaCharlie and I are building an app which uses Sarama to consume from a Kafka topic with the Consumer Group consumer. The relevant code looks like this:

```
	saramaClient, err := sarama.NewClient(consumerConfig.Brokers, consumerConfig.SaramaConfig)
	if err != nil {
		return nil, err
	}

	cGroup, err := sarama.NewConsumerGroupFromClient(consumerConfig.GroupId, saramaClient)
	if err != nil {
		return nil, err
	}

	handler := &seekBackHandler{
		client:         saramaClient,
		seekBackPeriod: cfg.SeekBackPeriod,
		registry:       consumerConfig.SchemaRegistry,
		consumer:       consumer,
	}

	go func() {
		// This needs to be invoked in a loop since cGroup.Consume returns any time a kafka rebalance occurs
		for {
			err := cGroup.Consume(ctx, cfg.ConsumerConfig.Topics, handler)
			if err != nil {
				slog.Error("Error during consumption, retrying", "err", err)
			}
			if ctx.Err() != nil {
				slog.Info("Context cancelled, exiting")
				return
			}
		}
	}()
```

One thing we are noticing is that if a single Kafka broker goes down transiently, this code will repeatedly retry to connect to the same Kafka broker rather than (as I expected) attempting to connect to different brokers if the first one is unavailable (see logs in the logs section)

Are we using Sarama correctly? I noticed there is a function `RefreshBrokers`, so one option would be to call `RefreshBrokers` when `cGroup.Consume` returns an error, but this seems expensive as it closes all existing broker connections, and I'm also not sure if it would work or if we'd also need to call `RefreshMetadata`.

Thanks in advance for your help!

##### Versions

| Sarama | Kafka | Go  |
| ------ | ----- | --- |
|  v1.38.1      |  not 100% sure, I think a fork off 3.0.0    |  1.22   |

##### Configuration

```go
    <property name="consumer.MaxProcessingTime" value="0" />
    <property name="consumer.GroupSessionTimeout" value="0" />
    <property name="consumer.GroupHeartbeatInterval" value="0" />
    <property name="consumer.InitTimeout" value="300" />
    <property name="consumer.InitialOffset" value="Newest" />
```

##### Logs

<details><summary>logs: CLICK ME</summary>
<p>

```
2024-07-31 22:15:47.929 ERROR Error during consumption, retrying err=dial tcp [2a04:f547:43:e046::11b3]:16637: connect: connection refused
2024-07-31 22:15:47.931 ERROR Error during consumption, retrying err=dial tcp [2a04:f547:43:e046::11b3]:16637: connect: connection refused
2024-07-31 22:15:47.933 ERROR Error during consumption, retrying err=dial tcp [2a04:f547:43:e046::11b3]:16637: connect: connection refused
2024-07-31 22:15:47.935 ERROR Error during consumption, retrying err=dial tcp [2a04:f547:43:e046::11b3]:16637: connect: connection refused

```

</p>
</details>

**Comment by ZoabKapoor** [comment]
Oh, one important thing to mention is that we're providing the `BrokerList` as a URL which our internal DNS then resolves into multiple hosts
```
<property name="consumer.BrokerList" value="kafka.<OURDOMAIN>.com:16637">
```

```
dig kafka.queuing.kafka.<OURDOMAIN>.com

;; ANSWER SECTION:
kafka.<OURDOMAIN>.com. 3 IN A 10.199.80.101
kafka.<OURDOMAIN>.com. 3 IN A 10.199.81.75
kafka.<OURDOMAIN>.com. 3 IN A 10.199.53.167
```

**Comment by thorseraq** [comment]
It may be DNS cache, every time it resolves to the same IP. if you provide only one broker address, (relying on dynamic ip resolving). Ideally, if you provide multiple address, sarama would init brokers with those address, and pick one from "LeastLoadedBroker", which is first previously connected broker and has least amount response promise to handle, if none, from unused seedBrokers(If conn failed, broker will be moved to deadSeeds)

**Comment by ZoabKapoor** [comment]
Interesting, DNS is definitely one of the things I'm considering, thanks @thorseraq ! Do you have any insight how I'd verify if this is the case/how to determine what values Sarama is getting out of the DNS cache? What's weird is that when I `dig AAAA` or `dig` the hostname, it correctly both (1) resolves and (2) shuffles the order each time, so I'm not sure why Sarama might be getting something different repeatedly

One thing we did do is hook up the Sarama logging, and we're hoping to see if this happens again if the Sarama-level logs can tell us a little more about what's going on.

**Comment by dnwe** [comment]
Hi @ZoabKapoor 

The key thing to remember with Kafka is that the broker connection addresses that you give to Sarama are simply what it can use to "bootstrap" a connection to the cluster. So by providing Sarama with `kafka.<OURDOMAIN>.com:<KAFKA_PORT>` you're telling it that there is one broker/hostname that can be used to find out where things live on the cluster.

The individual brokers themselves have a config for ADVERTISED_LISTENER which is the hostname:port that should be used to connect to the given individual broker. The log snippet above is very small, but do you know what configuration you are using for advertised listener on each broker?

**Comment by ZoabKapoor** [comment]
Hey @dnwe , that makes sense... I don't have direct access to the brokers to see what they're configuring as their `ADVERTISED_LISTENER`, unless that's something I can see in the sarama runtime using a debugger.

One thing that's interesting is that I've been trying to reproduce this issue locally and I'm unable to -- originally I thought that maybe what was happening is that the [dial](https://github.com/IBM/sarama/blob/a46c78d8078332141d3c7e9a6c6cc22e14af0f74/client.go#L1283) was returning the same value when resolving the DNS record `kafka.<OURDOMAIN>.com:<KAFKA_PORT>`, but testing it multiple times I could see that the IP returned varied, and to your point that only matters at bootstrap time -- once the client is directly connected to the leaders for the partitions it's consuming, it doesn't care about the initial bootstrap connection.

My guess as to what might have been happening (hard to verify because we can't reproduce) is the 7 seconds we saw many calls to connect to `2a04:f547:43:e046::11b`, the broker at `2a04:f547:43:e046::11b1` may have been down but the partition leader was not failed over yet... we should be able to determine if this happens again as we've enabled more detailed logging on the Sarama side.

For now, I will close this issue and reopen once we have more info (if it recurs) from the Sarama-level logs. 

Thanks to both of you for your help so far!
